### ISSUE TITLE
[ide-diagnostics] unused-variables false-positive with diagnostics.enableExperimental

### ISSUE THREAD

**Issue by bryceberger** [rust-lang/rust-analyzer#21549]
Possibly related to #20254

When `diagnostics.enableExperimental` is turned on, `bad` is diagnosed as an unused variable. It is not diagnosed as such without the feature, on stable, or by rustc.

**rust-analyzer version**: rust-analyzer 1.94.0-nightly (22c74ba 2026-01-15)

**rustc version**: rustc 1.94.0-nightly (22c74ba91 2026-01-15)

**editor or extension**: helix

**relevant settings**: `diagnostics.enableExperimental = true`

**code snippet to reproduce**:
```rust
fn main() {
    _ = |bad: u8| {
      // ^^^ unused-variable
        let inner = |i: &u8| *i == bad;
        [].iter().any(inner)
    };

    _ = |bad: u8| [].iter().any(|x: &u8| *x == bad);
    //   ^^^ unused-variable
}
```

```shell
$ rust-analyzer +1.93 diagnostics src/main.rs

diagnostic scan complete

$ rust-analyzer +nightly diagnostics src/main.rs
at crate playground, file /home/bryce/rust/playground/src/main.rs: Warning RustcLint("unused_variables") from LineCol { line: 1, col: 9 } to LineCol { line: 1, col: 12 }: unused variable
at crate playground, file /home/bryce/rust/playground/src/main.rs: Warning RustcLint("unused_variables") from LineCol { line: 7, col: 9 } to LineCol { line: 7, col: 12 }: unused variable
                                                                        
diagnostic scan complete
```

**Comment by ChayimFriedman2** [comment]
We disabled this diagnostic for this reason. This is not synced into nightly I believe.

**Comment by bryceberger** [comment]
Ah, presumably #21386 and #21445
